test/recipes/81-test_cmp_cli.t

```diff
@@ -0,0 +1,294 @@
+            $kur_port    = $1 eq "" ? '""""' : $1 if m/^\s*kur_port\s*=\s*(.*)?\s*$/;
+            $pbm_port    = $1 eq "" ? '""""' : $1 if m/^\s*pbm_port\s*=\s*(.*)?\s*$/;
+            $pbm_ref     = $1 eq "" ? '""""' : $1 if m/^\s*pbm_ref\s*=\s*(.*)?\s*$/;
+            $pbm_secret  = $1 eq "" ? '""""' : $1 if m/^\s*pbm_secret\s*=\s*(.*)?\s*$/;
+            $column      = $1 eq "" ? '""""' : $1 if m/^\s*column\s*=\s*(.*)?\s*$/;
+            $sleep       = $1 eq "" ? '""""' : $1 if m/^\s*sleep\s*=\s*(.*)?\s*$/;
+        }
+    }
+    close CH;
+    die "Cannot find all CMP server config values in $test_config section [$section]\n"
+        if !defined $ca_dn || !defined $server_dn
+        || !defined $server_cn || !defined $server_ip
+        || !defined $server_port || !defined $server_tls
+        || !defined $server_path || !defined $server_cert
+        || !defined $kur_port || !defined $pbm_port
+        || !defined $pbm_ref || !defined $pbm_secret
+        || !defined $column || !defined $sleep;
+    $server_dn = $server_dn // $ca_dn;
+}
+
+my @server_configurations = ("Mock");
+@server_configurations = split /\s+/, $ENV{CMP_SERVER} if $ENV{CMP_SERVER};
+# set env variable, e.g., CMP_SERVER="Mock Insta" to include further CMP servers
+
+my @all_aspects = ("connection", "verification", "credentials", "commands", "enrollment");
+@all_aspects = split /\s+/, $ENV{CMP_ASPECTS} if $ENV{CMP_ASPECTS};
+# set env variable, e.g., CMP_ASPECTS="commands enrollment" to select specific aspects
+
+my $faillog;
+if ($ENV{HARNESS_FAILLOG}) {
+    my $file = $ENV{HARNESS_FAILLOG};
+    open($faillog, ">", $file) or die "Cannot open $file for writing: $!";
+}
+
+sub test_cmp_cli {
+    my $server_name = shift;
+    my $aspect = shift;
+    my $n = shift;
+    my $i = shift;
+    my $title = shift;
+    my $params = shift;
+    my $expected_exit = shift;
+    my $path_app = bldtop_dir($app);
+    with({ exit_checker => sub {
+        my $actual_exit = shift;
+        my $OK = $actual_exit == $expected_exit;
+        if ($faillog && !$OK) {
+            sub quote_spc_empty(_) { $_ eq "" ? '""' : $_ =~ m/ / ? '"'.$_.'"' : $_ };
+            my $invocation = ("$path_app ").join(' ', map quote_spc_empty @$params);
+            print $faillog "$server_name $aspect \"$title\" ($i/$n)".
+                " expected=$expected_exit actual=$actual_exit\n";
+            print $faillog "$invocation\n\n";
+        }
+        return $OK; } },
+         sub { ok(run(cmd([$path_app, @$params,])),
+                  $title); });
+}
+
+sub test_cmp_cli_aspect {
+    my $server_name = shift;
+    my $aspect = shift;
+    my $tests = shift;
+    subtest "CMP app CLI $server_name $aspect\n" => sub {
+        my $n = scalar @$tests;
+        plan tests => $n;
+        my $i = 1;
+        foreach (@$tests) {
+          SKIP: {
+              test_cmp_cli($server_name, $aspect, $n, $i++, $$_[0], $$_[1], $$_[2]);
+              sleep($sleep);
```

Comment:
Why do we sleep here? I don't see anywhere that it is set to anything but 0

Reply:
The sleep value is not needed for local servers, but this script also for use with other servers.
For instance, the Insta Demo CA typically requires sleep = 3 or 4, otherwise it tends to return HTTP status 503.